File: app/src/main/java/io/github/golok56/manajemenuang/ui/transaction/form/TransactionFormActivity.java
package io.github.golok56.manajemenuang.ui.transaction.form;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.view.View;
import android.widget.EditText;
import android.widget.RadioButton;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.github.golok56.manajemenuang.R;
import io.github.golok56.manajemenuang.models.Transaction;
import io.github.golok56.manajemenuang.ui.base.BaseActivity;
import io.github.golok56.manajemenuang.ui.transaction.detail.TransactionDetailActivity;
import io.github.golok56.manajemenuang.utility.FormatUtil;
import io.github.golok56.manajemenuang.ui.main.MainActivity;

public class TransactionFormActivity extends BaseActivity {

    @BindView(R.id.et_amount_activity_transaction_form)
    EditText mEtAmount;
    @BindView(R.id.et_desc_activity_transaction_form)
    EditText mEtDescription;
    @BindView(R.id.rb_income_activity_transaction_form)
    RadioButton mCbIncome;
    private TransactionFormViewModel mViewModel;
    private boolean isEditMode = false;

    public static Intent getIntent(Context context) {
        return new Intent(context, TransactionFormActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_transaction_form);
        ButterKnife.bind(this);

        mViewModel = ViewModelProviders.of(this).get(TransactionFormViewModel.class);

        ActionBar ab = getSupportActionBar();
        if (ab != null) {
            ab.setDisplayHomeAsUpEnabled(true);
        }

        long id = getIntent().getLongExtra(TransactionDetailActivity.TRANSACTION_ID_EXTRA, -1);
        if (id != -1) {
            mViewModel.setId(id);
            mViewModel.getTransaction().observe(this, transaction -> {
                mEtAmount.setText(mViewModel.getAmount());
                mEtDescription.setText(mViewModel.getDesc());
                if (mViewModel.isIncome()) {
                    mCbIncome.setChecked(true);
                }
            });

            isEditMode = true;
        }

    }

    @OnClick(R.id.btn_save_activity_transaction_form)
    public void save(View v) {
        double amount;
        try {
            amount = Double.parseDouble(mEtAmount.getText().toString());
        } catch (Exception ex) {
            mEtAmount.setError(getString(R.string.validasi_form));
            return;
        }

        int type = mCbIncome.isChecked() ? Transaction.INCOME_TYPE : Transaction.EXPENSE_TYPE;
        String desc = mEtDescription.getText().toString();
        String date = FormatUtil.formatDate("dd MMMM yyyy");

        mViewModel.save(amount, type, desc, date);
        if (isEditMode) {
            showToast(getString(R.string.success_edit_transaction));
        } else {
            showToast(getString(R.string.success_create_transaction));
        }

        startActivity(MainActivity.getIntent(this));
    }

}

File: app/src/main/java/io/github/golok56/manajemenuang/ui/transaction/TransactionFormActivity.kt
package io.github.golok56.manajemenuang.ui.transaction

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.os.Bundle
import io.github.golok56.manajemenuang.R
import io.github.golok56.manajemenuang.models.Transaction
import io.github.golok56.manajemenuang.ui.base.BaseActivity
import io.github.golok56.manajemenuang.ui.main.MainActivity
import io.github.golok56.manajemenuang.utility.Formatter
import io.github.golok56.manajemenuang.viewmodels.TransactionFormViewModel
import kotlinx.android.synthetic.main.activity_transaction_form.*

class TransactionFormActivity : BaseActivity() {

    private lateinit var mViewModel: TransactionFormViewModel
    private var isEditMode = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_transaction_form)

        mViewModel = ViewModelProviders.of(this).get(TransactionFormViewModel::class.java)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val id = intent.getLongExtra(TransactionDetailActivity.TRANSACTION_ID_EXTRA, -1)
        if (id != -1L) {
            mViewModel.setId(id)
            mViewModel.transaction.observe(this, Observer { transaction ->
                etAmount.setText(Formatter.formatCurrency(transaction?.amount))
                etDesc.setText(transaction?.description)
                if (mViewModel.isIncome) {
                    rbIncome.isChecked = true
                }
            })

            isEditMode = true
        }

        btnSave.setOnClickListener{_ -> save()}
    }

    fun save() {
        val amount: Double
        try {
            amount = java.lang.Double.parseDouble(etAmount.text.toString())
        } catch (ex: Exception) {
            etAmount.error = getString(R.string.validasi_form)
            return
        }

        val type = if (rbIncome.isChecked) Transaction.INCOME_TYPE else Transaction.EXPENSE_TYPE
        val desc = etDesc.text.toString()
        val date = Formatter.formatDate("dd MMMM yyyy")

        mViewModel.save(amount, type, desc, date)
        if (isEditMode) {
            showToast(getString(R.string.success_edit_transaction))
        } else {
            showToast(getString(R.string.success_create_transaction))
        }

        start(MainActivity::class.java)
    }

    companion object {

        fun getIntent(context: Context): Intent {
            return Intent(context, TransactionFormActivity::class.java)
        }
    }

}
